packages/eslint-plugin/src/rules/no-confusing-void-expression.ts

```diff
@@ -376,5 +405,36 @@ export default createRule<Options, MessageId>({
       const type = getConstrainedTypeAtLocation(services, targetNode);
       return tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike);
     }
+
+    function targetNodeAncestorHasVoidReturnType(
+      node: TSESTree.Node,
+      targets: (
+        | AST_NODE_TYPES.FunctionDeclaration
+        | AST_NODE_TYPES.ArrowFunctionExpression
+        | AST_NODE_TYPES.FunctionExpression
+      )[],
+    ): boolean {
+      if (!node.parent) {
+        return false;
+      }
+
+      const filter = targets.filter(target => {
+        if (target === node.parent.type) {
+          if (
+            node.parent.returnType?.typeAnnotation.type ===
+            AST_NODE_TYPES.TSVoidKeyword
+          ) {
+            return true;
+          }
+        }
+        return false;
+      });
+
+      if (filter.length > 0) {
+        return true;
+      }
+
+      return targetNodeAncestorHasVoidReturnType(node.parent, targets);
+    }
```

Comment:
[Bug] This function won't stop until it finds a matching parent, so the following case is false negative

```ts
function foo(): void {
  const bar = () => {
    return console.log()
  }
}
```

([playground](https://deploy-preview-8632--typescript-eslint.netlify.app/play/#ts=5.3.3&fileType=.ts&code=GYVwdgxgLglg9mABMOcAUBKAXIgbnGAE0QG8AoRRCBAZykQCMBDAJ0QF5FMOA%2BUiyohYBTKCBZJqYGnAA2wgHSy4Ac0wCAvmS1A&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1rI6YDNYyZgHNaANw6UAJvQAexaCiG90YANrhsORNGgdokADSatWLdkiVhnRQDVJUgJJN701fjiIT2AL7eAupp%2BPkA&tsconfig=&tokens=false))

I think that this function should look +- like this:

- if `!node.parent` -> return false
- if `node.parent` is one of the `FunctionDeclaration`, `ArrowFunctionExpression`, `FunctionExpression`:
  - if `node.parent` has `void` in type annotation -> return true
  - if `node.parent` hasn't `void` in type annotation -> return false
- return `targetNodeAncestorHasVoidReturnType(node.parent)`

Reply:
in playground, this case seems false positive. anyway, this is bug, so i add test code and fix `targetNodeAncestorHasVoidReturnType` function

Reply:
I'm not familiar with the testing, so I think I used the terminology the other way around. thank you for telling me!